vortex-file/src/scan/mod.rs

```diff
@@ -0,0 +1,286 @@
+        self
+    }
+
+    pub fn with_task_executor(mut self, task_executor: TaskExecutor) -> Self {
+        self.task_executor = Some(task_executor);
+        self
+    }
+
+    pub fn with_metrics(mut self, metrics: VortexMetrics) -> Self {
+        self.metrics = metrics;
+        self
+    }
+
+    #[allow(clippy::unused_enumerate_index)]
+    pub fn build(self) -> VortexResult<impl ArrayStream + 'static> {
+        // Spin up the root layout reader, and wrap it in a FilterLayoutReader to perform
+        // conjunction splitting.
+        let layout_reader = Arc::new(FilterLayoutReader::new(self.vxf.layout_reader()?));
+
+        // Normalize and simplify the expressions.
+        let projection = simplify_typed(self.projection.clone(), layout_reader.dtype())?;
+        let filter = self
+            .filter
+            .clone()
+            .map(|f| simplify_typed(f, layout_reader.dtype()))
+            .transpose()?;
+
+        // Construct field masks and compute the row splits of the scan.
+        let (filter_mask, projection_mask) =
+            filter_and_projection_masks(&projection, filter.as_ref(), layout_reader.dtype())?;
+        let field_mask: Vec<_> = filter_mask
+            .iter()
+            .cloned()
+            .chain(projection_mask.iter().cloned())
+            .collect();
+        let splits = self.split_by.splits(layout_reader.layout(), &field_mask)?;
+        let row_indices = self.row_indices.clone();
+        let row_masks = splits
+            .into_iter()
+            .filter_map(move |row_range| {
+                let Some(row_indices) = &row_indices else {
+                    // If there is no row indices filter, then take the whole range
+                    return Some(RowMask::new_valid_between(row_range.start, row_range.end));
+                };
+
+                // Otherwise, find the indices that are within the row range.
+                let intersection = range_intersection(&row_range, row_indices)?;
+
+                // Construct a row mask for the range.
+                let filter_mask = Mask::from_indices(
+                    usize::try_from(row_range.end - row_range.start)
+                        .vortex_expect("Split ranges are within usize"),
+                    row_indices[intersection]
+                        .iter()
+                        .map(|&idx| {
+                            usize::try_from(idx - row_range.start)
+                                .vortex_expect("index within range")
+                        })
+                        .collect(),
+                );
+                Some(RowMask::new(filter_mask, row_range.start))
+            })
+            .collect_vec();
+
+        let result_dtype = projection.return_dtype(layout_reader.dtype())?;
+
+        // Create a future to process each row split of the scan.
+        let array_futures: Vec<_> = row_masks
+            .into_iter()
+            .enumerate()
+            .map(|(_i, row_mask)| {
```

Comment:
if its used, why the underscore?

Reply:
It's only used when tracing feature is enabled, so clippy shouts otherwise.